File: frontend/src/lib/reverseRequest.ts
import { api } from "./api";
import type { Page } from "./catalog";
import { qs } from "./qs";

export type ReverseRequestStatus = "PENDING_REVIEW" | "APPROVED" | "REJECTED" | "COMPLETED";
export type ReverseRequestType = "FULFILLMENT_CANCEL";
export type ReverseGoodsStatus = "IN_TRANSIT" | "RECEIVED";
export type ReverseSupplierResolution = "SUPPLIER_ACCEPTS_RETURN" | "COMPANY_BEAR_LOSS";

export interface ReverseRequestOut {
  id: number;
  no: string;
  request_type: ReverseRequestType;
  status: ReverseRequestStatus;
  sales_order_id: number;
  purchase_order_id: number;
  inbound_order_id: number;
  goods_status: ReverseGoodsStatus;
  supplier_resolution: ReverseSupplierResolution | null;
  reason: string;
  review_note: string | null;
  completion_note: string | null;
  created_at: string;
  updated_at: string;
  reviewed_at: string | null;
  completed_at: string | null;
  sales_order_no?: string | null;
  purchase_order_no?: string | null;
  inbound_order_no?: string | null;
  customer_display?: string | null;
  supplier_display?: string | null;
}

export interface ReverseRequestLineOut {
  id: number;
  reverse_request_id: number;
  inbound_order_line_id: number;
  purchase_order_line_id: number;
  sku_id: number;
  name_snapshot: string;
  spec_text_snapshot: string;
  unit_snapshot: string;
  qty: number | string;
}

export interface ReverseRequestListItem {
  id: number;
  no: string;
  request_type: ReverseRequestType;
  status: ReverseRequestStatus;
  sales_order_id: number;
  purchase_order_id: number;
  inbound_order_id: number;
  sales_order_no: string;
  purchase_order_no: string;
  inbound_order_no: string;
  customer_display: string;
  supplier_display: string;
  goods_status: ReverseGoodsStatus;
  supplier_resolution: ReverseSupplierResolution | null;
  line_count: number;
  total_qty: number | string;
  created_at: string;
}

export interface ReverseRequestDetail {
  request: ReverseRequestOut;
  lines: ReverseRequestLineOut[];
}

export interface ReverseRequestListFilters {
  status?: string;
  sales_order_id?: number;
  inbound_order_id?: number;
  q?: string;
  page?: number;
  size?: number;
}

export const reverseRequestApi = {
  list: (p: ReverseRequestListFilters) =>
    api.get<Page<ReverseRequestListItem>>(`/api/v1/reverse-requests${qs(p as Record<string, unknown>)}`),
  get: (id: number) => api.get<ReverseRequestDetail>(`/api/v1/reverse-requests/${id}`),
  create: (body: { inbound_order_id: number; reason: string }) =>
    api.post<ReverseRequestDetail>("/api/v1/reverse-requests", body),
  approve: (id: number, body: { supplier_resolution: ReverseSupplierResolution; review_note?: string | null }) =>
    api.post<ReverseRequestDetail>(`/api/v1/reverse-requests/${id}/approve`, body),
  reject: (id: number, body: { review_note?: string | null }) =>
    api.post<ReverseRequestDetail>(`/api/v1/reverse-requests/${id}/reject`, body),
  complete: (id: number, body: { completion_note?: string | null }) =>
    api.post<ReverseRequestDetail>(`/api/v1/reverse-requests/${id}/complete`, body),
};

File: backend/app/schemas/reverse_request.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class ReverseRequestCreateIn(BaseModel):
    inbound_order_id: int
    reason: str = Field(min_length=1, max_length=2000)


class ReverseRequestApproveIn(BaseModel):
    supplier_resolution: Literal["SUPPLIER_ACCEPTS_RETURN", "COMPANY_BEAR_LOSS"]
    review_note: str | None = Field(default=None, max_length=2000)


class ReverseRequestRejectIn(BaseModel):
    review_note: str | None = Field(default=None, max_length=2000)


class ReverseRequestCompleteIn(BaseModel):
    completion_note: str | None = Field(default=None, max_length=2000)


class ReverseRequestOut(BaseModel):
    id: int
    no: str
    request_type: str
    status: str
    sales_order_id: int
    purchase_order_id: int
    inbound_order_id: int
    goods_status: str
    supplier_resolution: str | None
    reason: str
    review_note: str | None
    completion_note: str | None
    created_at: datetime
    updated_at: datetime
    reviewed_at: datetime | None
    completed_at: datetime | None
    sales_order_no: str | None = None
    purchase_order_no: str | None = None
    inbound_order_no: str | None = None
    customer_display: str | None = None
    supplier_display: str | None = None

    @classmethod
    def build(cls, request, extra: dict | None = None) -> dict:
        d = cls.model_validate(request, from_attributes=True).model_dump()
        if extra:
            d.update(extra)
        return d


class ReverseRequestLineOut(BaseModel):
    id: int
    reverse_request_id: int
    inbound_order_line_id: int
    purchase_order_line_id: int
    sku_id: int
    name_snapshot: str
    spec_text_snapshot: str
    unit_snapshot: str
    qty: float

    @classmethod
    def build(cls, line) -> dict:
        return cls.model_validate(line, from_attributes=True).model_dump()


class ReverseRequestListItem(BaseModel):
    id: int
    no: str
    request_type: str
    status: str
    sales_order_id: int
    purchase_order_id: int
    inbound_order_id: int
    sales_order_no: str
    purchase_order_no: str
    inbound_order_no: str
    customer_display: str
    supplier_display: str
    goods_status: str
    supplier_resolution: str | None
    line_count: int
    total_qty: float
    created_at: datetime
